superbench/benchmarks/micro_benchmarks/ib_validation_performance/ib_validation_performance.cc

```diff
@@ -0,0 +1,450 @@
+            printf("Output will be saved to: %s\n", output_path.c_str());
+        }
+    }
+};
+
+// Joint vector<string> to vector<char> for mpi to send ips, each string split by '\0'.
+char *string_to_char(const std::vector<std::string> &strings) {
+    // the max length of each ip is 16
+    int max_len = 16;
+    // calculate size and allocate memory to save ips for all ranks
+    int size = (max_len + 1) * world_size + 1;
+    char *all_str = (char *)malloc(size * sizeof(char));
+    memset(all_str, '\0', size + 1);
+    // joint each string in the vector to a long char*
+    char *str_idx = all_str;
+    for (int i = 0; i < strings.size(); i++) {
+        memcpy(str_idx, strings[i].c_str(), (strings[i]).size());
+        str_idx = str_idx + max_len + 1;
+    }
+    return all_str;
+}
+
+// Split string by given symbol
+void split_string(const std::string &s, std::vector<std::string> &v, const std::string &c) {
+    std::string::size_type pos1, pos2;
+    pos2 = s.find(c);
+    pos1 = 0;
+    while (std::string::npos != pos2) {
+        if (pos1 != pos2) {
+            v.push_back(s.substr(pos1, pos2 - pos1));
+        }
+        pos1 = pos2 + c.size();
+        pos2 = s.find(c, pos1);
+    }
+    if (pos1 != s.length())
+        v.push_back(s.substr(pos1));
+}
+
+// Load and parse config file to vector
+vector<vector<std::pair<int, int>>> load_config(string filename = "config.txt") {
+    // read contents from file
+    vector<std::string> config;
+    ifstream in(filename);
+    string line;
+
+    if (in) {
+        while (getline(in, line)) {
+            config.push_back(line);
+        }
+    } else {
+        std::cout << "Error: failed to open config file." << std::endl;
+        MPI_Finalize();
+    }
+
+    if (world_rank == 0) {
+        std::cout << "config: " << std::endl;
+        for (const std::string &line : config) {
+            std::cout << line << std::endl;
+        }
+    }
+
+    // parse the string contents to vector
+    vector<vector<std::pair<int, int>>> run_in_total;
+    try {
+        for (const auto &single_line : config) {
+            // parse each line like "1,2;2,3;3,4" to vector<std::pair<int, int>>
+            vector<string> run_in_parallel;
+            vector<std::pair<int, int>> run_pairs_in_parallel;
+            // split line to pair by ";"
+            split_string(single_line, run_in_parallel, ";");
+            for (const auto &pair : run_in_parallel) {
+                // split pair by ","
+                int quote = pair.find(',');
+                int first = stoi(pair.substr(0, quote));
+                int second = stoi(pair.substr(quote + 1));
+                run_pairs_in_parallel.emplace_back(first, second);
+            }
+            run_in_total.emplace_back(run_pairs_in_parallel);
+        }
+    } catch (const std::exception &e) {
+        std::cout << "Error: failed to parse the config, message: " << e.what() << std::endl;
+        MPI_Finalize();
```

Comment:
in all the failure cases, please abort out. otherwise, the MPI library may hang (or unexpected behavior) as "not all processes will enter MPI_Finalize()"